src/Raven.Server/Documents/Revisions/RevisionsStorage.cs

```diff
@@ -868,54 +877,102 @@ public IEnumerable<string> GetCollections(Transaction transaction)
-            string changeVector, long lastModifiedTicks, 
+        private long DeleteRevisionsInternal(DocumentsOperationContext context, Table table, Slice lowerIdPrefix, CollectionName collectionName,
+            string changeVector, long lastModifiedTicks, long revisionsPreviousCount, 
             IEnumerable<Document> revisionsToRemove)
         {
             var writeTables = new Dictionary<string, Table>();
             long maxEtagDeleted = 0;
             var deleted = 0L;
 
+            var first = true;
+            Document lastRevisionToDelete = null;
+
             foreach (var revision in revisionsToRemove)
             {
-                using (Slice.From(context.Allocator, revision.ChangeVector, out var keySlice))
+                if (first)
                 {
-                    CreateTombstone(context, keySlice, revision.Etag, collectionName, changeVector, lastModifiedTicks);
+                    lastRevisionToDelete = revision;
+                    first = false;
+                    continue;
+                }
 
-                    maxEtagDeleted = Math.Max(maxEtagDeleted, revision.Etag);
-                    if (revision.Flags.Contain(DocumentFlags.HasAttachments))
-                    {
-                        _documentsStorage.AttachmentsStorage.DeleteRevisionAttachments(context, revision, changeVector, lastModifiedTicks);
-                    }
+                maxEtagDeleted = Math.Max(maxEtagDeleted, lastRevisionToDelete.Etag);
+                DeleteRevisionFromTable(context, table, writeTables, lastRevisionToDelete, collectionName, changeVector, lastModifiedTicks);
 
-                    Table writeTable = null;
-                    if (table.ReadByKey(keySlice, out TableValueReader tvr) && table.IsOwned(tvr.Id))
-                    {
-                        writeTable = table;
-                    }
-                    else
-                    {
-                        // We request to delete revision with the wrong collection
-                        var revisionData = TableValueToRevision(context, ref tvr, DocumentFields.Data);
+                deleted++;
+                lastRevisionToDelete = revision;
+            }
 
-                        var collection = _documentsStorage.ExtractCollectionName(context, revisionData.Data);
-                        if (writeTables.TryGetValue(collection.Name, out writeTable) == false)
-                        {
-                            writeTable = EnsureRevisionTableCreated(context.Transaction.InnerTransaction, collection);
-                            writeTables[collection.Name] = writeTable;
-                        }
-                    }
 
-                    writeTable.DeleteByKey(keySlice);
-                }
+            if (lastRevisionToDelete != null)
+            {
+                var remained = revisionsPreviousCount - deleted;
+                var skipLast = lastRevisionToDelete.Flags.Contain(DocumentFlags.DeleteRevision) && remained > 1 &&
```

Comment:
write a comment here, explaining this special case

Reply:
Done